packages/gitbook/src/components/DocumentView/CodeBlock/ClientCodeBlock.tsx

```diff
@@ -22,53 +20,56 @@ type ClientBlockProps = Pick<BlockProps<DocumentBlockCode>, 'block' | 'style'> &
 export function ClientCodeBlock(props: ClientBlockProps) {
     const { block, style, inlines } = props;
     const blockRef = useRef<HTMLDivElement>(null);
-    const processedRef = useRef(false);
-    const isInViewportRef = useRef<boolean | null>(null);
-    const [lines, setLines] = useState<HighlightLine[]>(() => plainHighlight(block, []));
+    const [isInViewport, setIsInViewport] = useState(false);
+    const plainLines = useMemo(() => plainHighlight(block, []), [block]);
+    const [lines, setLines] = useState<null | HighlightLine[]>(null);
 
     // Preload the highlighter when the block is mounted.
     useEffect(() => {
         import('./highlight').then(({ preloadHighlight }) => preloadHighlight(block));
     }, [block]);
 
-    const runHighlight = useEventCallback(() => {
-        if (processedRef.current) {
-            return;
-        }
-        if (typeof window !== 'undefined') {
-            import('./highlight').then(({ highlight }) => {
-                highlight(block, inlines).then((lines) => {
-                    setLines(lines);
-                    processedRef.current = true;
-                });
-            });
-        }
-    });
-    const debouncedRunHighlight = useDebounceCallback(runHighlight, 1000);
-
+    // Detect when the block is in viewport
     useInViewportListener(
         blockRef,
         (isInViewport, disconnect) => {
-            // Disconnect once in viewport
             if (isInViewport) {
+                // Disconnect once in viewport
                 disconnect();
-                // If it's initially in viewport, we need to run the highlight
-                if (isInViewportRef.current === null) {
-                    runHighlight();
-                }
+                setIsInViewport(true);
             }
-            isInViewportRef.current = isInViewport;
         },
         { rootMargin: '200px' }
     );
 
-    const handleScroll = useDebounceCallback(() => {
-        if (isInViewportRef.current) {
-            debouncedRunHighlight();
```

Comment:
@gregberge I don't understand what you tried to do here: if `isInViewportRef.current` is true, then you already ran the highlighting when it becomes visible, why defer another highlighting?

Reply:
I trigger highlighting on scroll and not from the observer, so it's smoother.

But the main issue on scroll is the outline, not that so I think you can simplify it.

Reply:
The main problem is triggering compute while the user scrolls. It's not good. I wanted to highlight when the user pause the scroll.

Reply:
But are you sure it's the issue? Because the blocks are empty if it was the issue they would be not highlighted but displayed at least.

Reply:
Because the initial presence in viewport was working so, I have some doubts on this fix.